File: smalivm/src/main/java/org/cf/smalivm/opcode/AGetOp.java
package org.cf.smalivm.opcode;

import org.cf.smalivm.ExceptionFactory;
import org.cf.smalivm.context.ExecutionNode;
import org.cf.smalivm.context.HeapItem;
import org.cf.smalivm.context.MethodState;
import org.cf.smalivm.dex.CommonTypes;
import org.jf.dexlib2.builder.MethodLocation;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.lang.reflect.Array;

public class AGetOp extends MethodStateOp {

    private static final Logger log = LoggerFactory.getLogger(AGetOp.class.getSimpleName());

    private final int valueRegister;
    private final int arrayRegister;
    private final int indexRegister;
    private final ExceptionFactory exceptionFactory;

    AGetOp(MethodLocation location, MethodLocation child, int valueRegister, int arrayRegister, int indexRegister,
           ExceptionFactory exceptionFactory) {
        super(location, child);
        this.valueRegister = valueRegister;
        this.arrayRegister = arrayRegister;
        this.indexRegister = indexRegister;
        this.exceptionFactory = exceptionFactory;

        addException(exceptionFactory.build(this, NullPointerException.class));
        addException(exceptionFactory.build(this, ArrayIndexOutOfBoundsException.class));
    }

    private static String getUnknownArrayInnerType(HeapItem array) {
        String outerType = array.getType();
        String result;
        if (CommonTypes.UNKNOWN.equals(outerType)) {
            result = CommonTypes.UNKNOWN;
        } else {
            result = outerType.replaceFirst("\\[", "");
        }

        return result;
    }

    @Override
    public void execute(ExecutionNode node, MethodState mState) {
        HeapItem arrayItem = mState.readRegister(arrayRegister);
        HeapItem indexItem = mState.readRegister(indexRegister);
        HeapItem getItem;
        if (arrayItem.isUnknown()) {
            String innerType = getUnknownArrayInnerType(arrayItem);
            getItem = HeapItem.newUnknown(innerType);
        } else {
            Object array = arrayItem.getValue();
            if (indexItem.isUnknown()) {
                String innerType = arrayItem.getType().replaceFirst("\\[", "");
                getItem = HeapItem.newUnknown(innerType);
            } else {
                // All values known, so exceptions are deterministic.
                node.clearExceptions();

                if (null == array) {
                    Throwable exception = exceptionFactory.build(this, NullPointerException.class);
                    node.setException(exception);
                    node.clearChildren();
                    return;
                }

                int index = indexItem.asInteger();
                String innerType = arrayItem.getType().replaceFirst("\\[", "");
                if (index >= Array.getLength(array)) {
                    Throwable exception = exceptionFactory.build(this, ArrayIndexOutOfBoundsException.class);
                    node.setException(exception);
                    node.clearChildren();
                    return;
                } else {
                    Object value = Array.get(array, index);
                    getItem = new HeapItem(value, innerType);
                    node.clearExceptions();
                }
            }
        }
        mState.assignRegister(valueRegister, getItem);
    }

    @Override
    public String toString() {
        return getName() + " r" + valueRegister + ", r" + arrayRegister + ", r" + indexRegister;
    }

}

File: smalivm/src/main/java/org/cf/smalivm/opcode/AGetOp.kt
package org.cf.smalivm.opcode

import ExceptionFactory
import org.cf.smalivm.configuration.Configuration
import org.cf.smalivm.dex.CommonTypes
import org.cf.smalivm.dex.SmaliClassLoader
import org.cf.smalivm.type.ClassManager
import org.cf.smalivm2.ExecutionNode
import org.cf.smalivm2.OpChild
import org.cf.smalivm2.Value
import org.cf.util.Utils
import org.jf.dexlib2.builder.MethodLocation
import org.jf.dexlib2.iface.instruction.formats.Instruction23x
import org.slf4j.LoggerFactory
import java.lang.reflect.Array

class AGetOp internal constructor(
    location: MethodLocation,
    child: MethodLocation,
    private val destRegister: Int,
    private val arrayRegister: Int,
    private val indexRegister: Int,
) : Op(location, child, NullPointerException::class.java, ArrayIndexOutOfBoundsException::class.java) {

    override val registersReadCount = 2
    override val registersAssignedCount = 2

    override fun execute(node: ExecutionNode): kotlin.Array<out OpChild> {
        val array = node.state.readRegister(arrayRegister)
        val index = node.state.readRegister(indexRegister)
        val indexed: Value

        var mayThrow = true
        if (array.isUnknown) {
            val innerType = getUnknownArrayInnerType(array)
            indexed = Value.unknown(innerType)
        } else {
            if (index.isUnknown) {
                val innerType = array.type.replaceFirst("\\[".toRegex(), "")
                indexed = Value.unknown(innerType)
            } else {
                if (array.isNull) {
                    return throwChild(NullPointerException::class.java)
                }
                val innerType = array.type.replaceFirst("\\[".toRegex(), "")
                if (index.toInteger() >= Array.getLength(array.value)) {
                    return throwChild(ArrayIndexOutOfBoundsException::class.java)
                } else {
                    val raw = Array.get(array.value, index.toInteger())
                    indexed = Value.wrap(raw, innerType)
                    mayThrow = false
                }
            }
        }
        node.state.assignRegister(destRegister, indexed)
        return collectChildren(mayThrow)
    }

    override fun toString() = "$name r$destRegister, r$arrayRegister, r$indexRegister"

    companion object : OpFactory {
        private val log = LoggerFactory.getLogger(AGetOp::class.java.simpleName)

        private fun getUnknownArrayInnerType(array: Value): String {
            val outerType = array.type
            return if (CommonTypes.UNKNOWN == outerType) {
                CommonTypes.UNKNOWN
            } else {
                outerType.replaceFirst("\\[".toRegex(), "")
            }
        }

        override fun build(
            location: MethodLocation,
            addressToLocation: Map<Int, MethodLocation>,
            classManager: ClassManager,
            exceptionFactory: ExceptionFactory,
            classLoader: SmaliClassLoader,
            configuration: Configuration
        ): Op {
            val child = Utils.getNextLocation(location, addressToLocation)
            val instr = location.instruction as Instruction23x
            val valueRegister = instr.registerA
            val arrayRegister = instr.registerB
            val indexRegister = instr.registerC
            return AGetOp(location, child, valueRegister, arrayRegister, indexRegister)
        }
    }
}
